src/CodeGen_X86.cpp

```diff
@@ -600,6 +617,24 @@ void CodeGen_X86::visit(const Call *op) {
         }
     }
 
+    // Check for saturating_pmulhrs. On x86, pmulhrs is truncating, but it's still faster
+    // to use pmulhrs than to lower (producing widening multiplication), and have a check
+    // for the singular overflow case.
+    static Expr saturating_pmulhrs = rounding_mul_shift_right(wild_i16x_, wild_i16x_, 15);
+    if (expr_match(saturating_pmulhrs, op, matches)) {
+        // Rewrite so that we can take advantage of pmulhrs.
+        internal_assert(matches.size() == 2);
+        const Expr &a = matches[0];
+        const Expr &b = matches[1];
+        Expr pmulhrs = i16(rounding_shift_right(widening_mul(a, b), 15));
+        // Handle edge case of possible overflow.
+        Expr i16_min = op->type.min();
+        Expr i16_max = op->type.max();
+        Expr expr = select((a == i16_min) && (b == i16_min), i16_max, pmulhrs);
```

Comment:
On AVX2 or <32 vector lanes, the following is better (assuming it's correct)
```
Expr mask = select(max(a, b) == i16_min), cast<int16_t>(-1), cast<int16_t>(0));
Expr expr = mask ^ pmulhrs;
```
This should be four instructions (vpmaxsw, vcmpeqw, vpmulhrsw, vpxor)

The original is five (vcmpeqw, vcmpeqw, vpand, vpmulhrs, vpclendvb)

For avx512 and vectorized at least 32-wide, it seems like what you already have is better, because the comparison result goes into a mask register.

Reply:
That rewrite formally verifies for me. I will add it!

Reply:
Added in [41f09f8](https://github.com/halide/Halide/pull/7129/commits/41f09f83f7e108aa4856401db09a24d6ec559a70)